### DISCUSSION TITLE
Control Schemes and Customization

### DISCUSSION THREAD

**Discussion by ViMaSter** [Studio-Lovelies/GG-JointJustice-Unity#237] [Ideas]
Based on the conversation that took place around controls, resulting in https://github.com/Studio-Lovelies/GG-JointJustice-Unity/issues/229, this thread should be used to discuss **things we want for the control scheme of the game** while taking into consideration **when they should be implemented**.

**For each new idea, make a new post in this thread**, so the discussion can continue on sub-threads for that specific feature. (Should you come up with a new one while discussing another, still feel free to make a new post for it, so each feature has a dedicated thread and maybe link to the discussion up to this point.)

**Comment by ViMaSter** [comment]
Being part of a country (unfortunately) consisting mainly of `QWERTZ` keyboards, keyboard control schemes that use `Y`, `X`, and `C` are fairly unintuitive to use. (Imagine if you'd start a game that maps their buttons to `Z`, `X`, `C`.)

I see a quick and easy step to mitigate this, which **I would love to see being integrated for the Case 1 Unity rerelease already**:
Unity's new input system allows us to find the [the display value of the specific keycode](https://docs.unity3d.com/Packages/com.unity.inputsystem@0.9/api/UnityEngine.InputSystem.InputControl.html#:~:text=Note%20that%20the%20display%20name%20of%20a%20control%20may%20change%20over%20time.%20For%20example%2C%20when%20changing%20from%20a%20QWERTY%20keyboard%20layout%20to%20an%20AZERTY%20keyboard%20layout%2C%20the%20%22q%22%20key%20(which%20will%20keep%20that%20name)%20will%20change%20its%20display%20name%20from%20%22q%22%20to%20%22a%22.).  
For reference, this is what happens if the Z key is pressed on a German keyboard layout in the browser **(notice the value of `code`)**:
![image](https://user-images.githubusercontent.com/1689033/160493039-d0de5189-08a8-4cac-8194-98ef0bc09ab9.png)

As soon as our in-game control displays are dynamic (so we don't have to use sprites to show the controls at the button of a menu), we could use this to update the keys for specific locales.

**Reply by PhoebeMitchell** [reply]
Just to clarify what the problem is here:
The location of the Z is different on some keyboards so a Z/X/C control scheme is difficult to use?

From what I understand Input System already accounts for this. It identifies keys based on their physical location rather than the letter on them.

If the problem is that the controls in the menus don't update to reflect this then hell yeah, we should make that change!

**Reply by ViMaSter** [reply]
> Just to clarify what the problem is here:
The location of the Z is different on some keyboards so a Z/X/C control scheme is difficult to use?
From what I understand Input System already accounts for this. It identifies keys based on their physical location rather than the letter on them.

Just tried this out in-game and, yup, that actually already works. Didn't know that, wild assumption, my bad!

> If the problem is that the controls in the menus don't update to reflect this then hell yeah, we should make that change!

_That_ however is actually still is a problem, you're right! The menu keeps showing Z, even though the Y button is at that position when I switch keyboard layouts with Alt+Shift.

**Comment by ViMaSter** [comment]
I'm going to preface this by saying that I rarely play text adventures; let alone on PC, so do take this with a grain of salt:  
I find [the default keyboard control scheme for Ren'Py (a Python text adventure engine) games](https://defkey.com/renpy-shortcuts) (like Doki Doki Literature Club) to be fairly intuitive; more so than our current `Y`/`X`/`C` control scheme. (As it would also solve [localization issues with non-QWERTY keyboards](https://github.com/Studio-Lovelies/GG-JointJustice-Unity/discussions/237#discussioncomment-2456596).)

In my opinion, this is not urgent at all, so **I would be fine if this is done after Case 1**, but maybe changing schemes mid-releases is even more irritating, than just changing it before that.

**Reply by PhoebeMitchell** [reply]
Personally, as someone who grew up playing questionably obtained Pokemon roms on emulators where A was bound to X and B was bound to Z, Z/X/C comes perfectly naturally to me.

That's a bit of an obscure influence, but the game _is_ supposed to be based on GameBoy Advance Ace Attorney games, and more modern games like Undertale use a similar control scheme.

Also, it probably differs between keyboards but personally find spacebar to be a very "clunky" key.

**Comment by PhoebeMitchell** [comment]
I feel like it's impossible to please everyone in this scenario without having some form of customisable controls. We chose to use Input System to be able to do this, and while changing controls isn't entirely intuitive it's not exactly impossible. Making the UI for it would probably be harder.
